### PR TITLE
IGNITE-18222: Implement the rest of dynamic completors

### PR DESCRIPTION
**Pull request by Flaugh24**
https://issues.apache.org/jira/browse/IGNITE-18222
Added completers for script-file, cluster-url, jdbc-url option

### CODE REVIEW

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/completer/path/FilePathDynamicCompleterFactory.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+package org.apache.ignite.internal.cli.core.repl.completer.path;
+
+import jakarta.inject.Singleton;
+import org.apache.ignite.internal.cli.core.repl.completer.DynamicCompleter;
+import org.apache.ignite.internal.cli.core.repl.completer.DynamicCompleterFactory;
+
+/** Factory for file path option completer. */
+@Singleton
+public class FilePathDynamicCompleterFactory implements DynamicCompleterFactory {
```
Comment (PakhomovAlexander) [inline modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/completer/path/FilePathDynamicCompleterFactory.java:26]: I think if the implementation of completer factory is so simple that can be a lambda – it should be a lambda.

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/executor/ReplExecutorProvider.java [hunk 2]
```diff
@@ -35,7 +35,7 @@ public class ReplExecutorProvider {
     private Terminal terminal;
 
     @Inject
-    private NodeNameRegistry nodeNameRegistry;
+    private NodeNameRegistryImpl nodeNameRegistry;
```
Comment (PakhomovAlexander) [inline modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/executor/ReplExecutorProvider.java:38]: There are multiple places where you've renamed the `NodeNameRegistry` to `NodeNameRegistryImpl` I think this is done by mistake. Could you fix all of them?

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/registry/impl/JdbcUrlRegistryImpl.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+import org.apache.ignite.internal.thread.NamedThreadFactory;
+import org.apache.ignite.rest.client.api.NodeConfigurationApi;
+import org.apache.ignite.rest.client.invoker.ApiException;
+import org.apache.ignite.rest.client.invoker.Configuration;
+
+/** Implementation of {@link JdbcUrlRegistry}. */
+@Singleton
+public class JdbcUrlRegistryImpl implements JdbcUrlRegistry, SessionEventListener {
+
+    private final IgniteLogger log = Loggers.forClass(getClass());
```
Comment (PakhomovAlexander) [inline modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/registry/impl/JdbcUrlRegistryImpl.java:46]: CliLoggers

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/registry/impl/JdbcUrlRegistryImpl.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    private final NodeNameRegistryImpl nodeNameRegistry;
+    private volatile Set<JdbcUrl> jdbcUrls = Set.of();
```
Comment (PakhomovAlexander) [inline modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/registry/impl/JdbcUrlRegistryImpl.java:48]: I don't think the `volatile` is the best choice here. Maybe `ConcurrentHashSet`?

Comment (PakhomovAlexander) [inline modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/registry/impl/JdbcUrlRegistryImpl.java:48]: `JdbcUrl` should implement `equals` and `hashcode`.

**Review by PakhomovAlexander — CHANGES_REQUESTED** [review]

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/registry/impl/JdbcUrlRegistryImpl.java (hunk 4 above)
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/registry/impl/JdbcUrlRegistryImpl.java:48]: Why? Right now we don't remove/add anything from/to the map.

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/JdbcUrl.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        this.host = host;
+        this.port = port;
+    }
+
+    /** Returns {@link JdbcUrl} with provided host and port. */
+    public static JdbcUrl of(String url, int port) throws MalformedURLException {
+        return new JdbcUrl(new URL(url).getHost(), port);
+    }
+
+    public String toString() {
```
[the remark this answers is not in the document]
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/JdbcUrl.java:41]: done

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/Session.java [hunk 6]
```diff
@@ -18,51 +18,60 @@
...
-        return nodeName;
+    public Session(List<SessionEventListener> listeners) {
+        this.listeners = listeners;
     }
 
-    public void setNodeName(String nodeName) {
-        this.nodeName = nodeName;
+    /** Creates session details with provided nodeUrl, nodeName, jdbcUrl. */
+    public synchronized void connect(String nodeUrl, String nodeName, String jdbcUrl) {
+        this.sessionDetails = new SessionDetails(nodeUrl, nodeName, jdbcUrl);
```
[the remark this answers is not in the document]
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/Session.java]: done

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/Session.java [hunk 7]
```diff
@@ -18,51 +18,60 @@
...
 
-    public void setNodeUrl(String nodeUrl) {
-        this.nodeUrl = nodeUrl;
+    /** Clears session details and sets false to connectedToNode. */
+    public synchronized void disconnect() {
+        this.sessionDetails = new SessionDetails();
+        this.connectedToNode = false;
+        listeners.forEach(it -> {
+            try {
+                it.onConnect(this);
```
[the remark this answers is not in the document]
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/Session.java]: done

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/executor/ReplExecutorProvider.java (hunk 2 above)
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/executor/ReplExecutorProvider.java]: done

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/registry/impl/JdbcUrlRegistryImpl.java (hunk 4 above)
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/registry/impl/JdbcUrlRegistryImpl.java:51]: done

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/registry/impl/JdbcUrlRegistryImpl.java (hunk 3 above)
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/registry/impl/JdbcUrlRegistryImpl.java]: done

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/completer/path/FilePathDynamicCompleterFactory.java (hunk 1 above)
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/completer/path/FilePathDynamicCompleterFactory.java]: done

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/Session.java [hunk 8]
```diff
@@ -18,51 +18,60 @@
...
-    public String jdbcUrl() {
-        return jdbcUrl;
+    public boolean isConnectedToNode() {
+        return connectedToNode;
     }
 
-    public void setJdbcUrl(String jdbcUrl) {
-        this.jdbcUrl = jdbcUrl;
+    /** Returns {@link SessionDetails}. */
+    public SessionDetails sessionDetails() {
```
[the remark this answers is not in the document]
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/Session.java]: done

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/SessionEventListener.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.cli.core.repl;
+
+/** Session event listener. */
+public interface SessionEventListener {
```
[the remark this answers is not in the document]
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/SessionEventListener.java]: done

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/Session.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/Session.java]: done

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/Session.java [hunk 10]
```diff
@@ -18,51 +18,60 @@
...
+                it.onConnect(this);
+            } catch (Exception e) {
+                log.warn("Got an exception: ", e);
+            }
+        });
     }
 
-    public String jdbcUrl() {
-        return jdbcUrl;
+    public boolean isConnectedToNode() {
```
[the remark this answers is not in the document]
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/Session.java:69]: done

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/completer/DynamicCompleterActivationPoint.java [hunk 11]
```diff
@@ -86,5 +97,28 @@ public void activateDynamicCompleter(DynamicCompleterRegistry registry) {
...
+                CompleterConf.builder()
+                        .command("sql")
+                        .enableOptions(Options.JDBC_URL)
+                        .exclusiveEnableOptions().build(),
+                jdbcUrlDynamicCompleterFactory
+        );
+
+        registry.register(
+                CompleterConf.builder()
+                        .enableOptions(Options.CLUSTER_URL)
```
[the remark this answers is not in the document]
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/completer/DynamicCompleterActivationPoint.java]: done

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/Session.java [hunk 12]
```diff
@@ -18,51 +18,60 @@
...
  */
 @Singleton
 public class Session {
 
-    private boolean connectedToNode;
-
-    private String nodeUrl;
-
-    private String nodeName;
+    private final IgniteLogger log = Loggers.forClass(getClass());
```
[the remark this answers is not in the document]
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/Session.java]: done

**Comment by Pochatkin** [comment]
LGTM

**Pull request merged**